### PR TITLE
Move task ordinal inheritance later during graph calculation

### LINKED ISSUE #20195: A task that is finalized by a task specified earlier on the command line causes the build to hang
For example, if I have a task `generateFoo` that is finalized by `cleanupFoo` but then I specify `cleanupFoo generateFoo` on the command line, then I end up with a situation where `cleanUpFoo` must run after `generateFoo`, but `generateFoo` cannot run because the dependencies of `cleanupFoo` (i.e. `generateFoo`) have not completed.  The command line ordering rules conflict with the finalization relationship and cause the build to hang.

### Expected Behavior
This conflict should be handled gracefully such that the finalization is honored, even if the command line order is not.

### Current Behavior
The build hangs.

### PR DESCRIPTION
**Pull request by ghale**
Moves the inheritence of ordinal value later in task graph calculation in order to capture all relationships derived through finalizers.

Fixes #20195

### CODE REVIEW

**Comment by ghale** [comment]
@bot-gradle test rFM

**Comment by ghale** [comment]
@bot-gradle test rFM

**Comment by ghale** [comment]
@bot-gradle test rFM

File: subprojects/core/src/integTest/groovy/org/gradle/api/DestroyerTaskCommandLineOrderIntegrationTest.groovy [hunk 1]
```diff
@@ -310,4 +311,62 @@ class DestroyerTaskCommandLineOrderIntegrationTest extends AbstractCommandLineOr
...
+        def cleanBar = bar.task('cleanBar').destroys('build/bar').dependsOn(cleanFoo)
+        def clean = rootBuild.task('clean').dependsOn(cleanFoo).dependsOn(cleanBar)
+        def generateFoo = foo.task('generateFoo').produces('build/foo', type).finalizedBy(cleanFoo)
+        def generateBar = bar.task('generateBar').produces('build/bar', type)
+        def generate = rootBuild.task('generate').dependsOn(generateBar).dependsOn(generateFoo)
+
+        writeAllFiles()
+
+        when:
+        executer.startBuildProcessInDebugger(true)
```
Comment (big-guy) [inline subprojects/core/src/integTest/groovy/org/gradle/api/DestroyerTaskCommandLineOrderIntegrationTest.groovy:359]: ditto

Comment (big-guy) [inline subprojects/core/src/integTest/groovy/org/gradle/api/DestroyerTaskCommandLineOrderIntegrationTest.groovy:359]: ditto

**Review by big-guy — APPROVED** [review]

File: subprojects/core/src/integTest/groovy/org/gradle/api/DestroyerTaskCommandLineOrderIntegrationTest.groovy [hunk 2]
```diff
@@ -310,4 +311,62 @@ class DestroyerTaskCommandLineOrderIntegrationTest extends AbstractCommandLineOr
...
+        def cleanBar = bar.task('cleanBar').destroys('build/bar')
+        def clean = rootBuild.task('clean').dependsOn(cleanFoo).dependsOn(cleanBar)
+        def generateFoo = foo.task('generateFoo').produces('build/foo', type).finalizedBy(cleanFoo)
+        def generateBar = bar.task('generateBar').produces('build/bar', type)
+        def generate = rootBuild.task('generate').dependsOn(generateBar).dependsOn(generateFoo)
+
+        writeAllFiles()
+
+        when:
+        executer.startBuildProcessInDebugger(true)
```
[the remark this answers is not in the document]
Comment (ghale) [reply subprojects/core/src/integTest/groovy/org/gradle/api/DestroyerTaskCommandLineOrderIntegrationTest.groovy]: Bah, yes.  Just forgot to remove it.  Gets me every time.

**Comment by ghale** [comment]
@bot-gradle test rFM

**Comment by ghale** [comment]
@bot-gradle test and merge

**Pull request merged**